reconcile/jenkins_job_builds_cleaner.py

```diff
@@ -0,0 +1,78 @@
+import logging
+import re
+import time
+
+import reconcile.queries as queries
+
+from reconcile.utils.jenkins_api import JenkinsApi
+
+QONTRACT_INTEGRATION = 'jenkins-job-builds-cleaner'
+
+
+def hours_to_ms(hours):
+    return hours * 60 * 60 * 1000
+
+
+def run(dry_run):
+    jenkins_instances = queries.get_jenkins_instances()
+    settings = queries.get_app_interface_settings()
+
+    time_ms = time.time() * 1000
+
+    for instance in jenkins_instances:
+        instance_cleanup_rules = instance.get('buildsCleanupRules', [])
+        if not instance_cleanup_rules:
+            # Skip instance if no cleanup rules defined
+            continue
+
+        # Process cleanup rules, pre-compile as regexes
+        cleanup_rules = []
+        for rule in instance_cleanup_rules:
+            cleanup_rules.append({
+                'name': rule['name'],
+                'name_re': re.compile(rule['name']),
+                'keep_hours': rule['keep_hours'],
+                'keep_ms': hours_to_ms(rule['keep_hours'])
+            })
+
+        instance_name = instance['name']
+        token = instance['token']
+        jenkins = JenkinsApi(token, ssl_verify=False, settings=settings)
+        all_job_names = jenkins.get_job_names()
+
+        delete_builds = []
+
+        for job_name in all_job_names:
+            builds = None
+            for rule in cleanup_rules:
+                if rule['name_re'].search(job_name):
+                    # Fetch list of builds if we dont have it already
+                    if not builds:
+                        builds = jenkins.get_builds(job_name)
+                    for build in builds:
+                        build_id = build['id']
+                        if time_ms - rule['keep_ms'] > build['timestamp']:
+                            delete_builds.append({
+                                'job_name': job_name,
+                                'rule_name': rule['name'],
+                                'rule_keep_hours': rule['keep_hours'],
+                                'build_id': build['id'],
+                            })
+
+        todel_current = 0
+        todel_total = len(delete_builds)
+        for build in delete_builds:
+            todel_current += 1
+            job_name = build['job_name']
+            build_id = build['build_id']
+            progress_str = f"{todel_current}/{todel_total}"
+            logging.info(['clean_job_builds', progress_str,
+                         instance_name, job_name, build['rule_name'],
+                         build['rule_keep_hours'], build_id])
+            if not dry_run:
+                try:
+                    jenkins.delete_build(build['job_name'], build['build_id'])
+                except Exception as e:
```

Comment:
Can you capture a specific exception class? I've suffered way more pain than I want to remember with code that captured `Exception`.

Reply:
I tried to see what we could do without refactoring too much of the jenkins_api library. 

At the moment the delete_build method (most methods from jenkins_api.py really) will simply return `res.raise_for_status()` from `requests`. As best I can tell I would need to catch an exception from `requests` here but this code isn't aware of requests. I could import it, but really `requests` is an implementation detail of the library.

I think the best solution would be to refactor the jenkins_api.py library and have all methods return proper library-specific exceptions instead of `requests`.

Does that make sense? Can you suggest something I may have missed?